libcst/_nodes/statement.py

```diff
@@ -2017,13 +2017,30 @@ class With(BaseCompoundStatement):
     #: Sequence of empty lines appearing before this with statement.
     leading_lines: Sequence[EmptyLine] = ()
 
+    #: Optional open parenthesis for multi-line with bindings
+    lpar: Union[LeftParen, MaybeSentinel] = MaybeSentinel.DEFAULT
+
+    #: Optional close parenthesis for multi-line with bindings
+    rpar: Union[RightParen, MaybeSentinel] = MaybeSentinel.DEFAULT
```

Comment:
If we use `MaybeSentinel` here instead of `Optional[LeftParen]`, we need to be very clever during code generation.
The main difference between these options is that `Optional` signals that the caller is responsible for determining if a `LeftParen` is needed or not (and `With` may or may not validate this). If we go with `MaybeSentinel`, then `With` needs to make sure the parenthesis are emitted when necessary (i.e. if the `with` statement's header spans multiple lines).

The second option is a better UX overall if we can come up with the cleverness to detect when the parens are necessary (which would be nice to do in `_validate_parens` even if we go with `Optional`).

Reply:
Makes sense.

It seems like the rust side doesn't care which we do, so I think I'll try to get the parser working first. I think it will be easier to figure out the right approach once I can generate the CST from an example.

Reply:
>  the rust side doesn't care

Absolutely. Ergonomy of the rust API is a tertiary priority (if that) 😄

Reply:
Hoping to do this tomorrow, for both the rust and python codegen steps.

I *think* that I probably only need to look for whitespace on the commas of the WithItem entries in `items` to make this work properly.